### PR TITLE
S0071 masked verhoeff

### PR DESCRIPTION
**Pull request by KnowledgeForwardSolutions**
This pull request adds masked validation support to the Verhoeff algorithm, along with comprehensive unit tests, readme updates and benchmark results. Also corrects an oversight in earlier masked validation changes by adding explicit tests for a null check digit mask as well as unit tests for that case.

### CODE REVIEW

File: CheckDigits.Net/GeneralAlgorithms/VerhoeffAlgorithm.cs [hunk 1]
```diff
@@ -93,4 +93,56 @@ public Boolean Validate(String value)
...
+         if (n.IsInvalidDigit())
+         {
+            return false;
+         }
+
+         p = _permutationTable[i % 8, n];
+         c = _multiplicationTable[c, p];
+
+         i++;
+         processedDigits ++;
```
[the remark this answers is not in the document]
Comment (KnowledgeForwardSolutions) [reply CheckDigits.Net/GeneralAlgorithms/VerhoeffAlgorithm.cs]: fixed

File: CheckDigits.Net.Tests.Unit/GeneralAlgorithms/Modulus11_27ExtendedAlgorithmTests.cs [hunk 2]
```diff
@@ -247,6 +247,15 @@ public void Modulus11_27ExtendedAlgorithm_Validate_ShouldReturnTrue_ForBenchmark
    // ==========================================================================
    // ==========================================================================
 
+   [Fact]
+   public void  Modulus11_27ExtendedAlgorithm_ValidateMasked_ShouldThrowArgumentNullException_WhenMaskIsNull()
```
[the remark this answers is not in the document]
Comment (KnowledgeForwardSolutions) [reply CheckDigits.Net.Tests.Unit/GeneralAlgorithms/Modulus11_27ExtendedAlgorithmTests.cs]: fixed

File: CheckDigits.Net.Tests.Benchmarks/NumericAlgorithmBenchmarks.cs [hunk 3]
```diff
@@ -274,12 +282,12 @@ public void TryCalculateCheckDigit(ISingleCheckDigitAlgorithm algorithm, String
...
-   public void Validate(ICheckDigitAlgorithm algorithm, String name, String value)
-   {
-      algorithm.Validate(value);
-   }
+   //[Benchmark]
+   //[ArgumentsSource(nameof(ValidateArguments))]
+   //public void Validate(ICheckDigitAlgorithm algorithm, String name, String value)
+   //{
+   //   algorithm.Validate(value);
+   //}
```
[the remark this answers is not in the document]
Comment (KnowledgeForwardSolutions) [reply CheckDigits.Net.Tests.Benchmarks/NumericAlgorithmBenchmarks.cs:290]: ignoring

File: CheckDigits.Net/IMaskedCheckDigitAlgorithm.cs [hunk 4]
```diff
@@ -31,5 +31,8 @@ public interface IMaskedCheckDigitAlgorithm : ICheckDigitAlgorithm
    ///   <see cref="String.Empty"/> or a string that is of invalid length for 
    ///   this algorithm.
    /// </remarks>
+   /// <exception cref="ArgumentNullException">
+   ///   <paramref name="mask"/> is <see langword="null"/>.
+   /// </exception>
```
[the remark this answers is not in the document]
Comment (KnowledgeForwardSolutions) [reply CheckDigits.Net/IMaskedCheckDigitAlgorithm.cs:36]: fixed

[the remark this answers is not in the document]
Comment (KnowledgeForwardSolutions) [reply CheckDigits.Net/IMaskedCheckDigitAlgorithm.cs:36]: fixed
